rfc/006-semantic-versioning.md

```diff
@@ -0,0 +1,84 @@
+2. **Node Operators**, those running node infrastructure, are predominantly concerned with downtime, complexity and frequency of upgrading and loss of data. They may be also concerned about changes that may break the scripts and tooling they use to supervise their nodes.
+3. **External Clients**, for example, wallets and block explorers are those that perform any of the following:
+     - consume the RPC endpoints of nodes like `/block`
+     - subscribe to events that are streamed via websockets
+     - make queries to the indexer
+
+    This set are concerned with chain upgrades which will impact their ability to query state and block data as well as broadcast transactions.
+
+4. There is also a fourth emerging user: **IBC module and relayers** (although in many ways relayers can also be considered sub components to external clients). The developers of IBC and consumers of their software are concerned about changes that may affect a chains ability to send arbitrary messages to another chain. Specifically, this can happen with a breaking change to the light client verification algorithm.  
+
+In a more broader sense, these four users are inextricably linked and their concerns reflect that of the end user of the applications. The major conclusion that this RFC makes is that **the ability for chains to provide continual service is more important than the actual upgrade burden put on the developers of these chains**. What that means is that although it may be burdensome for application developers to upgrade Tendermint versions and this may cause longer release cycles, the other three groups experience upgrades as downtime which ultimately affects end users of the application.
+
+### Modes of Interprocess Communication
+
+Tendermint has two primary mediums with which it can communicate to other processes: RPC and P2P. The division marks the boundary between the internal and external components of the network. The P2P layer is used in all cases that nodes (validator, full, seed or light) need to communicate with one another whereas the RPC is for any outside process that wants to communicate with a node. The assumption made here is that all communication via the RPC is to a trusted source and thus the ability to interpret information is more important than being able to verify the information. P2P is therefore primary medium for verification. As an example, an in-browser light client would consist of verifying headers (and perhaps application state) via the p2p layer and passing the information on to the client via the RPC layer (or potentially directly via an API).
+
+The exception to this is the IBC module and relayers which are external but require verifiable data. Breaking changes to the light client verification path mean that all neighbouring chains that are connected will no longer be able to verify state transitions and thus pass messages back and forward.
```

Comment:
I'm still uncertain as to what guarantees we should give this user group? Should we not break light client verification across the entire major version? 

The most problematic aspect about this is that individual nodes are easier to upgrade than entire networks. It may take 6 months for a network to update their IBC module and include it in the next release. I've been pondering on this for a while and I can't think of any tricks to get around this. ABCI++ will add vote extensions which will change the bytes that get signed over such that previous clients won't be able to verify new signatures. A couple ideas I've had to mitigate the damage. 
- Allow wasm-like clients that can be upgraded separately to the entire chain
- If there are signature changes, have validators sign over the current and previous data types ? Or have ways we can parse the old data formats from the new.

Other than that the best I can think of is to release the light client stuff before everything else so we can give the IBC team and their users as much time as possible to upgrade

 Thoughts? @creachadair @colin-axner

Reply:
> Should we not break light client verification across the entire major version?

Yes I think so (at least from the perspective of IBC and at least in the short term). Breaking the light client verification means that all counterparties must upgrade their code in order to prevent their IBC connection from breaking. 

As you say, using Wasm to upgrade clients as opposed to chain code upgrades can ease this burden, but I don't think we can assume every chain using IBC is running Wasm. 

> If there are signature changes, have validators sign over the current and previous data types ? Or have ways we can parse the old data formats from the new.

I think exploration of this idea is useful, but this would be a non-breaking solution (since it allows for backwards compatibility)

---

Does there exist light client breaking changes which would not normally cause a major version bump?

In general I'm still fairly concerned about upgrading IBC clients. I think the practical timeline of expecting counterparties to upgrade to support the new IBC client is minimum 1 year. Furthermore, even if most chains use Wasm, the current clients that have been initialized do not currently have a way to migrate to become Wasm clients

I don't see a way around this though since breaking the light client algorithm means new code needs to be added to handle the breakage. The only things that come to mind are trying to minimize the surface area of changes which require this upgrade path, either with backwards compatible fixes, wasm clients or hub routing (hub routing would only be beneficial if the hub is running a tm chain that will never break the light client algorithm or it has a wasm client and most connected chains have a wasm client).

Reply:
It's an interesting game-theory-like ploy to use breaking IBC as a forcing function to get networks to upgrade Tendermint versions 😄.

Anyway, I'am also concerned about this. I think even if it was to be a month of incompatibility it would still be disastrous for the ecosystem. Osmosis' entire value proposition rests on a functioning IBC. 

> Does there exist light client breaking changes which would not normally cause a major version bump?

Prior to this discussion most light client breaking changes would not result in a major version bump. A major version bump really means you need to start a new chain; your existing chain is incompatible with the new software. We want to reduce that as much as possible so all nodes (including light clients) will be able to verify data for all minor versions that are less than equal to their minor version. This means that all nodes need to upgrade ahead of when the entire network itself decides to change versions (In fact, validators won't vote to upgrade the network version if they don't have that software so you'll need 2/3+). I think this works fine within the boundaries of a single network which releases the same software for everyone but falls apart across multiple networks where each ecosystem has their own release schedules. 

Having wasm clients or some other method for dynamically upgrading clients would be a great thing to have in the future. We don't want to have too many light client breaking changes but they will probably still happen so putting something in place that can minimize the downtime of IBC when it happens is definitely worth investing in.

Reply:
I don't have a strong opinion here, but I guess for me the question that matters is this: Is there some reason we should _not_ guarantee against light-client breaking changes within a major release? 

As with the other user groups, I'm receptive to an escape hatch for security issues. Is there some other obvious technical reason why we might want to break light client verification specifically, without breaking other users of the protocol?

If we needed to do that anyway, then an argument to the incentive might be a reasonable justification for it—but if we don't have a clear need my intuition is we should include these users in the version guarantees.

Reply:
> Is there some other obvious technical reason why we might want to break light client verification specifically

This will quite often be a by-product of adding some other feature. Like PBTS could break the light client although we've chosen not to for 0.36. More direct forms of breaking the verification algorithm would be to change the way we do signatures, change what's in those signatures or change what's in the header.

Reply:
Understood. But wouldn’t those example changes also break other users as well?

Reply:
Yes but for node operators and external clients we find such breakage acceptable. Node operators need to upgrade their binary but this can be done automatically and networks should only upgrade when all/most of the validators have the next version running. For external clients we care more for readability than verifiability

Reply:
```suggestion
The main exceptions to this are the IBC module and relayers, which are external to the node but also require verifiable data. Breaking changes to the light client verification path mean that all neighbouring chains that are connected will no longer be able to verify state transitions and thus pass messages back and forward.
```

Reply:
If we agree that the IBC use case is high priority, perhaps we should start with a conservative guarantee for those users specifically.

Reply:
This [ADR](https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-027-ibc-wasm.md) might help. Wasm allows light clients to be upgraded by submitting a message to a chain. Existing light clients require the entire chain to upgrade to upgrade the light client